lib/netplay/tcp/poll_descriptor_set.h

```diff
@@ -0,0 +1,93 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+
+/*
+	This file is part of Warzone 2100.
+	Copyright (C) 2025  Warzone 2100 Project
+
+	Warzone 2100 is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+
+	Warzone 2100 is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with Warzone 2100; if not, write to the Free Software
+	Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
+*/
+
+#pragma once
+
+#include "descriptor_set.h"
+
+#include "lib/framework/frame.h" // for MAX_PLAYERS
+
+#ifdef WZ_OS_WIN
+# include <winsock2.h>
+#elif defined(WZ_OS_UNIX)
+# include <poll.h> // for pollfd, poll
+#endif
+
+#include <array>
+#include <stdexcept>
+
+namespace tcp
+{
+
+/// <summary>
+/// Descriptor set interface specialization using the `poll()` API for actual polling.
+/// </summary>
+/// <typeparam name="EventType">Type of updates (readable/writable sockets) to poll for.</typeparam>
+template <PollEventType EventType>
+class PollDescriptorSet : public IDescriptorSet
+{
+public:
+
+	explicit PollDescriptorSet() = default;
+
+	virtual bool add(SOCKET fd) override
+	{
+		constexpr short evt = EventType == PollEventType::READABLE ? POLLIN : POLLOUT;
+
+		assert(size_ < MAX_PLAYERS);
+		if (size_ >= MAX_PLAYERS)
+		{
+			return false;
+		}
+		fds_[size_++] = { fd, evt, 0 };
+		return true;
+	}
+
+	virtual void clear() override
+	{
+		fds_.fill({});
+		size_ = 0;
+	}
+
+	virtual int pollImpl(std::chrono::milliseconds timeout) override
+	{
+#ifdef WZ_OS_WIN
+		return WSAPoll(fds_.data(), size_, timeout.count());
+#else
+		return poll(fds_.data(), size_, timeout.count());
+#endif
+	}
+
+	virtual bool isSet(SOCKET fd) const override
+	{
+		constexpr short evt = EventType == PollEventType::READABLE ? POLLIN : POLLOUT;
+
+		const auto it = std::find_if(fds_.begin(), fds_.end(), [fd](const pollfd& pfd) { return pfd.fd == fd; });
+		return it != fds_.end() && (it->revents & evt);
+	}
+
+private:
+
+	std::array<pollfd, MAX_PLAYERS> fds_;
```

Comment:
`MAX_PLAYERS` is going to be insufficient:
- The host can currently have `MAX_TMP_SOCKETS` sockets in `tmp_socket`, which are all added to `tmp_socket_set`.
- The host can currently have `MAX_CONNECTED_PLAYERS` sockets in `connected_bsocket`, which are all added to `server_socket_set`.

(One or both of these may be increased in the future.)

Seems like we could make this a `std::vector<pollfd>`?

Reply:
Good catch! Increased limit to `MAX_CONNECTED_PLAYERS + MAX_TMP_SOCKETS`.